Asciidoctor to ReST doc conversion

gonsie opened the pull request:
Start of conversion for core documentation.

I’ll need some help integrating sphinx and such into the build system... all I’ve done thus far is remove the asciidoctor checks.

grondo commented:
Thanks! I can try to help integrate sphinx or whatever is required on the build system side of things.

You will also need to remove `doc/man*/Makefile` from `configure.ac` along with the `No asciidoctor formatter` warning (or we can do that. We still may want to allow the rest of flux-core to build without generated docs, and keep the warning) Though I'm not sure how the generated man pages will get installed at `make install` time without Makefiles?

Did COPYRIGHT inclusion get dropped on purpose in man pages? (I notice some other includes were preserved)

SteVwonder commented on doc/conf.py:
```diff
@@ -76,5 +76,227 @@
 
 # -- Options for man output -------------------------------------------------
 
+# Run the gen-manlines.sh script to generate the following text
```
Is `gen-manlines.sh` something you wrote or an external project?  If the former, should it be included with this commit?  If the latter, can a link to it be added to this comment?

gonsie replied:
I was originally going to include the script, but then I decided not to... do you think a script to generate these lines would be useful? Or, since man pages are added a few at a time, just have people write the lines by hand?

grondo replied:
Anything that can automate a manual process is good IME, since every manual process will generate an error at some point.

Also if the file is fully generated by a script, the script could be included in version control but not the result.

gonsie replied:
Since these lines are only generated once, I’m not inclined to put the generating script in. Really, the script is more a part of the conversation from adoc, rather than something that will be used repeatedly.

grondo replied:
Ah, ok, if the file never needs to be touched again, no reason to include the script!

gonsie commented:
@grondo This PR definitely was a rough start and I wasn’t sure the exact use-case that would need to be supported. I think the basic use case is: If the ‘enable docs’ option is on, then the docs should be built and installed with `make install`. Are there any others?

The copyright statement at the bottom of the man pages is generated & inserted by sphinx. Actually, the author statement is inserted as well, I’ll go and remove those.

gonsie commented:
One other use case is that the man pages get generated for the readthedocs website (duh!). Any others?

SteVwonder commented:
> One other use case is that the man pages get generated for the readthedocs website (duh!). Any others?

As a small addendum to this use case, I think it would be cool if we could have the RTD CI run on flux-core PRs like it does for the rfc and docs repos.  I'm not sure if we just need `make html` to pass through to sphinx, or if there is something more going on in the RTD CI.  We probably need a requirements.txt if we don't already have one.

gonsie commented:
@SteVwonder the travis.yaml file is super complex here... any idea where to start? maybe we can get together on friday to hack it out. :\

gonsie commented:
Question: which NAME section style do you prefer? Style 1 includes the other command that share the same man page text, style 2 just has the current man page name. The original style (included) is not possible with sphinx.

```
# Style 1
NAME
       flux-getattr - flux-getattr, flux-setattr, flux-lsattr - access broker attributes

# Style 2
NAME
       flux-getattr - access broker attributes

# Original
NAME
       flux-getattr, flux-setattr, flux-lsattr - access broker attributes some junk
```

SteVwonder commented:
> Question: which NAME section style do you prefer?

What would the NAME section for the `flux-lsattr` manpage be in style 2?
```
NAME
    flux-lsattr - access broker attributes
```
OR
```
NAME
    flux-getattr - access broker attributes
```

gonsie commented:
For both styles, the `flux-lsattr` name would come first on that man page. 

Basically the Sphinx generation puts “name - description” on this line, it just depends if we include the other related commands in the “description” text.

grondo commented:
> One other use case is that the man pages get generated for the readthedocs website (duh!). Any others?

I can't think of any. A finer point: the man pages should be generated with `make` and installed with `make install`.  If you run `src/cmd/flux help command|function` in the build dir, the MANPATH is constructed such that you should get the manpage from the build tree (i.e. the manpage that goes with the version of `flux` you are running)

SteVwonder commented:
> For both styles, the flux-lsattr name would come first on that man page.

Awesome! In that case, my vote is for style 2 since it is cleaner.

grondo commented:
We should try to get this merged asap, so that we won't have to keep rebasing it as new or changed documentation lands in flux-core.

Is there an option in the conversion process from adoc->rst to keep the line breaks from the source material? At least the files I spot checked, paragraphs end up all on one line. This might not be conducive to reviewing 
future changes, since a single change in a paragraph will create a diff that involves the whole single-line paragraph.

I'll try to take a look at getting sphinx run (when found) with `make`, and the docs installed with `make install`. I wonder if sphinx is already available in the TOSS buildfarm. (I hope so)

SteVwonder commented:
> I'll try to take a look at getting sphinx run (when found) with make, and the docs installed with make install.

Sorry for not posting a commit yesterday.  I have a half-way baked `Makefile.am` in the works from the doc day call yesterday. I'll try and get it working in the next hour or so.

It's pretty much a copy paste from the old `Makefile.am` that used asciidoc, except it is lifted up one directory and has `adoc` replaced for `rst`. 

> getting sphinx run (when found)

As long as we are OK requiring that the python that flux-core is configured against is the same python that has the sphinx module, I think we can treat it exactly like the other python dependencies.  I was thinking we can use the `AM_CHECK_PYMOD` m4 macro to check the module version is > 1.6.7 (where 1.6.7 is the version in the bionic package repo).  And we shouldn't need to locate the `sphinx_build` since it can be run via the configured python interpreter with `$(PYTHON) -m sphinx SPHINX_ARGS`.

I just gotta figure out the right automake syntax that supports "recursive" commands that can generate everything in one go, as opposed to compilers/asciidoc that can be run with one input file and produce one output file.

grondo commented:
> I just gotta figure out the right automake syntax that supports "recursive" commands that can generate everything in one go, as opposed to compilers/asciidoc that can be run with one input file and produce one output file.

This is the main issue I was anticipating. Thanks a lot for taking this on!! I didn't have time to look at it yet, but will happily test :stuck_out_tongue_winking_eye:

SteVwonder commented:
Just pushed a commit to a branch on my fork: https://github.com/SteVwonder/flux-core/commits/rst-docs

It works for generating the manpages.  I haven't tested `make install` or `make distcheck`.  Calling it for the weekend.  I can revisit next week unless someone else wants to push it forward.  If run `make` in the `doc` directory, it will run sphinx multiple times because it is looking for manpages for `idset_*`, which somehow got dropped in this conversion.  If you comment out those manpages in the `doc/Makefile.am`, then sphinx only gets run once.  The trick was this target:

```
$(MAN_FILES): $(RST_FILES)
    $(AM_V_GEN) $(PYTHON) -m sphinx -M man ./ ./
```

gonsie commented:
@SteVwonder @grondo I think this branch is working!

SteVwonder commented:
Thanks @gonsie for all the work on this PR! You are awesome! :1st_place_medal: 

---

I am getting an error when building the whole project.  The `gen-cmdhelp.pl` script seems to have references to `.adoc`.  I'm not sure exactly what is going on in the script (it looks like it outputs a json mapping of commands->manpages similar to what is in conf.py). Maybe @garlick or @grondo can comment on exactly what is happening, and if it is even necessary anymore since Sphinx generates a full manpage for each target rather than just so links.

```
  CC       flux-queue.o
  CCLD     flux-queue
  CC       flux-exec.o
  CCLD     flux-exec
make[3]: Leaving directory '/data/hdd/Repositories/flux-framework/flux-core/src/cmd'
make[2]: Leaving directory '/data/hdd/Repositories/flux-framework/flux-core/src/cmd'
Making all in test
make[2]: Entering directory '/data/hdd/Repositories/flux-framework/flux-core/src/test'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/data/hdd/Repositories/flux-framework/flux-core/src/test'
make[2]: Entering directory '/data/hdd/Repositories/flux-framework/flux-core/src'
make[2]: Nothing to be done for 'all-am'.
make[2]: Leaving directory '/data/hdd/Repositories/flux-framework/flux-core/src'
make[1]: Leaving directory '/data/hdd/Repositories/flux-framework/flux-core/src'
Making all in etc
make[1]: Entering directory '/data/hdd/Repositories/flux-framework/flux-core/etc'
  GEN      flux/help.d/core.json
Usage: ./gen-cmdhelp.pl --category=[label] FILES...
make[1]: *** [Makefile:830: flux/help.d/core.json] Error 255
make[1]: Leaving directory '/data/hdd/Repositories/flux-framework/flux-core/etc'
make: *** [Makefile:574: all-recursive] Error 1
```

---

All sorts of weirdness happens in `doc` when run under `make -j` (sorry for not catching that earlier).  Using `.NOTPARALLEL` seems to fix the issue:

```diff
diff --git a/doc/Makefile.am b/doc/Makefile.am
index 466747c1f..3bb863d1b 100644
--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -1,3 +1,5 @@
+.NOTPARALLEL:
+
 SUBDIRS = . test

 MAN_FILES = $(MAN1_FILES) $(MAN3_FILES) $(MAN5_FILES) $(MAN7_FILES)
```

---

It looks like configure didn't find `aspell`.  Maybe the `aspell` code snippet in `configure.ac` needs to get added back in?

```
make[2]: Leaving directory '/data/hdd/Repositories/flux-framework/flux-core/doc'
make[1]: Leaving directory '/data/hdd/Repositories/flux-framework/flux-core/doc'
Making check in test
make[1]: Entering directory '/data/hdd/Repositories/flux-framework/flux-core/doc/test'
make  check-TESTS
make[2]: Entering directory '/data/hdd/Repositories/flux-framework/flux-core/doc/test'
make[3]: Entering directory '/data/hdd/Repositories/flux-framework/flux-core/doc/test'
SKIP: spellcheck - skip because aspell is not installed
============================================================================
Testsuite summary for flux-core 0.17.0-101-g5f61facdd
============================================================================
# TOTAL: 1
# PASS:  0
# SKIP:  1
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
make[3]: Leaving directory '/data/hdd/Repositories/flux-framework/flux-core/doc/test'
make[2]: Leaving directory '/data/hdd/Repositories/flux-framework/flux-core/doc/test'
make[1]: Leaving directory '/data/hdd/Repositories/flux-framework/flux-core/doc/test'

sherbein ~/Repositories/flux-framework/flux-core/doc (rst-docs !?S)
❯ which aspell                                                                                                              14:49:43 ()
/usr/bin/aspell

(1) sherbein ~/Repositories/flux-framework/flux-core (rst-docs !?S)
❯ grep aspell configure.ac config.log || echo "no matches found"                                                                         14:51:09 ()
no matches found
```

gonsie commented:
Looks like `gen-cmdhelp.pl` processes comments from the adoc files (which got dropped in the conversion). It’d be helpful to see the results of the script, so I can figure out if there is an easy way to recreate them with sphinx.

grondo commented:
> Looks like gen-cmdhelp.pl processes comments from the adoc files (which got dropped in the conversion). It’d be helpful to see the results of the script, so I can figure out if there is an easy way to recreate them with sphinx.

Yeah, that script processes comments in command man pages to determine if commands should be added to the generic `flux help` summary of "commonly used commands".

```json
[
  {
    "category": "core",
    "command": "broker",
    "description": "Invoke Flux comms message broker daemon"
  },
  {
    "category": "core",
    "command": "content",
    "description": "Access instance content storage"
  },
  {
    "category": "core",
    "command": "cron",
    "description": "Schedule tasks on timers and events"
  },
  {
    "category": "core",
    "command": "dmesg",
    "description": "manipulate broker log ring buffer"
  },
  {
    "category": "core",
    "command": "env",
    "description": "Print or run inside a Flux environment"
  },
  {
    "category": "core",
    "command": "event",
    "description": "Send and receive Flux events"
  },
  {
    "category": "core",
    "command": "exec",
    "description": "Execute processes across flux ranks"
  },
  {
    "category": "core",
    "command": "get,set,lsattr",
    "description": "Access, modify, and list broker attributes"
  },
  {
    "category": "core",
    "command": "hwloc",
    "description": "Control/query resource-hwloc service"
  },
  {
    "category": "core",
    "command": "jobs",
    "description": "list jobs submitted to Flux"
  },
  {
    "category": "core",
    "command": "keygen",
    "description": "generate keys for Flux security"
  },
  {
    "category": "core",
    "command": "kvs",
    "description": "Flux key-value store utility"
  },
  {
    "category": "core",
    "command": "logger",
    "description": "create a Flux log entry"
  },
  {
    "category": "core",
    "command": "mini",
    "description": "Minimal Job Submission Tool"
  },
  {
    "category": "core",
    "command": "module",
    "description": "manage Flux extension modules"
  },
  {
    "category": "core",
    "command": "ping",
    "description": "measure round-trip latency to Flux services"
  },
  {
    "category": "core",
    "command": "proxy",
    "description": "Create proxy environment for Flux instance"
  },
  {
    "category": "core",
    "command": "start",
    "description": "bootstrap a local Flux instance"
  },
  {
    "category": "core",
    "command": "version",
    "description": "Display flux version information"
  }
]
```

grondo commented:
Can we just add the comments back to the source material?

gonsie commented:
@grondo I added back the comments, but there is something buggy about `gen-cmdhelp.pl` script. When I run it I get the error:
```
Use of uninitialized value $h{"command"} in printf at ./gen-cmdhelp.pl line 68
```
Can you check out the last commit I pushed (`7ea6860`) and see if you can figure out if anything is wrong?

grondo commented:
I ran into an issue testing this out in a VPATH build. (sphinx SRCDIR was listed as `.` instead of `$(srcdir)`). Also, I noticed `conf.py` wasn't added to `Makefile.am`, so this file would likely not be included in `make distcheck`.

Finally, "quiet mode" make wasn't working because sphinx spews a lot of gak by default.

And lastly, some manpages were mentioned twice and this generated automake warnings.

This patch should fix all that  up:

```diff
diff --git a/doc/Makefile.am b/doc/Makefile.am
index 3bb863d1b..41fa5c2fa 100644
--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -196,7 +196,6 @@ MAN3_FILES_SECONDARY = \
        man/flux_future_wait_for.3 \
        man/flux_future_reset.3 \
        man/flux_future_destroy.3 \
-       man/flux_future_get.3 \
        man/flux_future_fulfill.3 \
        man/flux_future_fulfill_error.3 \
        man/flux_future_fulfill_with.3 \
@@ -204,13 +203,11 @@ MAN3_FILES_SECONDARY = \
        man/flux_future_aux_get.3 \
        man/flux_future_set_flux.3 \
        man/flux_future_get_flux.3 \
-       man/flux_future_wait_all_create.3 \
        man/flux_future_wait_any_create.3 \
        man/flux_future_push.3 \
        man/flux_future_first_child.3 \
        man/flux_future_next_child.3 \
        man/flux_future_get_child.3 \
-       man/flux_future_and_then.3
        man/flux_future_or_then.3 \
        man/flux_future_continue.3 \
        man/flux_future_continue_error.3 \
@@ -282,10 +279,19 @@ endif
 
 SUFFIXES = .rst .1 .3 .5 .7
 
-$(MAN_FILES): $(RST_FILES)
-       $(AM_V_GEN) $(PYTHON) -m sphinx -M man ./ ./
+sphinx_verbose = $(sphinx_verbose_$(V))
+sphinx_verbose_ = $(sphinx_verbose_$(AM_DEFAULT_VERBOSITY))
+sphinx_verbose_0 = @echo "  SPHINX  manpages";
+
+STDERR_DEVNULL = $(stderr_devnull_$(V))
+stderr_devnull_ =  $(stderr_devnull_$(AM_DEFAULT_VERBOSITY))
+stderr_devnull_0 = >/dev/null 2>&1
+
+$(MAN_FILES): conf.py $(RST_FILES)
+       $(sphinx_verbose)$(PYTHON) -m sphinx -M man $(srcdir) ./ $(STDERR_DEVNULL)
 
 EXTRA_DIST = \
+       conf.py \
        $(RST_FILES) \
        man1/NODESET.rst \
        man3/JSON_PACK.rst \
```

I changed the quiet output to just `SPHINX manpages` instead of `SPHINX man1/flux.1`:
I also added `conf.py` as manpage generation dependency.

```Making all in .
make[1]: Entering directory '/home/grondo/git/f.git/build/doc'
  SPHINX  manpages
make[1]: Leaving directory '/home/grondo/git/f.git/build/doc'
Making all in test
make[1]: Entering directory '/home/grondo/git/f.git/build/doc/test'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/grondo/git/f.git/build/doc/test'


```

grondo commented:
> @grondo I added back the comments, but there is something buggy about gen-cmdhelp.pl script. When I run it I get the error:

Sorry @gonsie, I had somehow missed this comment. I apologize!

grondo commented:
Another change that breaks one of the current manpage use cases:

When run in a builddir, `src/cmd/flux` sets `MANPATH` such that `man flux` will display the current builddir version of `flux.1`. This is done so that documentation matches the version of Flux being run.

However, sphinx places all output into a single directory `doc/man` instead of generating output in the section directories in which the man sources are found. This doesn't conform to a standard `MANPATH` setup, so the scheme described above fails.

```console
ƒ(s=1,d=0,builddir) grondo@asp:~/git/f.git$ echo $MANPATH
/home/grondo/git/f.git/build/doc:
ƒ(s=1,d=0,builddir) grondo@asp:~/git/f.git$ man -w flux
/usr/local/man/man1/flux.1
$ find . -name flux.1
./doc/man/flux.1
```

We should output man files in to the correct directories during build, but I can't figure out how to do this with sphinx (or even if it is possible).

SteVwonder commented:
> We should output man files in to the correct directories during build, but I can't figure out how to do this with sphinx (or even if it is possible).

I don't think it is possible "natively".  It looks like the class structure for the Sphinx builders is configured with a single outdir, and the target filename has the section included but the directory path does not: [[sphinx man builder source code]](https://github.com/sphinx-doc/sphinx/blob/af62fa61e6cbd88d0798963211e73e5ba0d55e6d/sphinx/builders/manpage.py#L85)

One thought would be to invoke sphinx once for each man directory and manually set the output directory to man1, man3, etc from the automake level.  My only worry with that is that sphinx will try and add an additional man, so that we would end up with `$FLUX_SRC/doc/man/man1/man/flux.1` rather than `$FLUX_SRC/doc/man/man1/flux.1`.  Maybe we can force it to not add that extra `man` directory though.

A janky workaround could be to use symlinks `ln -s ./man ./man/man1`.  Gross and it would break the `-S` functionality of `man`, but I think it would work.

Thanks @gonsie for opening the issue on sphinx's repo.

grondo commented:
>One thought would be to invoke sphinx once for each man directory and manually set the output directory to man1, man3, etc from the automake level. My only worry with that is that sphinx will try and add an additional man, so that we would end up with $FLUX_SRC/doc/man/man1/man/flux.1 rather than $FLUX_SRC/doc/man/man1/flux.1.

Thanks, I tried something like this and it came out pretty much as you described. I think to do it "correctly" you'd have to split the `conf.py` into a different one for each of the man sections (but I'm really just guessing. sphinx has too much magic for me to understand what it is doing)

I'm not opposed to symlinks for the sake of the builddir. It is a bit disappointing, but I think getting this PR merged sooner rather than later has more benefits than the drawback of symlinks. I can take a look at it tomorrow if nobody beats me to it.

gonsie commented:
@grondo @SteVwonder : 
Instead of symlinking, why not create the man1, man3 directories and move the files to the correct location directly in the makefile? 

```
$(MAN_FILES): $(RST_FILES)
    $(AM_V_GEN) $(PYTHON) -m sphinx -M man ./ ./
    mkdir man/man1 man/man3 man/man5 man/man7
    mv man/*.1 man/man1/
    ...
```

I still don't fully understand the makefiles though, so I'm not sure if this would muck up the install step.

grondo commented:
> Instead of symlinking, why not create the man1, man3 directories and move the files to the correct location directly in the makefile?

Not a bad idea! At first I thought this would result in the rule be re-run each time you run `make`, since the files are being moved. However, if we do it all under one rule as you have it might work!

grondo commented:
> Instead of symlinking, why not create the man1, man3 directories and move the files to the correct location directly in the makefile?

Ok, got this working. At first I was worried that moving sphinx output away from where it expects it would result in sphinx regenerating all manpages whenever the source of one changes, but it appears that sphinx does this anyway! 

Here's the (large) diff I ended up with after all the changes. With this applied, my testcase from above works, and also the build tree passes `make distcheck`!


```
diff --git a/doc/Makefile.am b/doc/Makefile.am
index 3bb863d1b..bd9978727 100644
--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -11,267 +11,265 @@ MAN_FILES = $(MAN1_FILES) $(MAN3_FILES) $(MAN5_FILES) $(MAN7_FILES)
 MAN1_FILES = $(MAN1_FILES_PRIMARY) $(MAN1_FILES_SECONDARY)
 
 MAN1_FILES_PRIMARY = \
-	man/flux.1 \
-	man/flux-broker.1 \
-	man/flux-kvs.1 \
-	man/flux-keygen.1 \
-	man/flux-logger.1 \
-	man/flux-ping.1 \
-	man/flux-start.1 \
-	man/flux-module.1 \
-	man/flux-exec.1 \
-	man/flux-env.1 \
-	man/flux-getattr.1 \
-	man/flux-dmesg.1 \
-	man/flux-content.1 \
-	man/flux-hwloc.1 \
-	man/flux-proxy.1 \
-	man/flux-cron.1 \
-	man/flux-event.1 \
-	man/flux-mini.1 \
-	man/flux-version.1 \
-	man/flux-jobs.1
+	man1/flux.1 \
+	man1/flux-broker.1 \
+	man1/flux-kvs.1 \
+	man1/flux-keygen.1 \
+	man1/flux-logger.1 \
+	man1/flux-ping.1 \
+	man1/flux-start.1 \
+	man1/flux-module.1 \
+	man1/flux-exec.1 \
+	man1/flux-env.1 \
+	man1/flux-getattr.1 \
+	man1/flux-dmesg.1 \
+	man1/flux-content.1 \
+	man1/flux-hwloc.1 \
+	man1/flux-proxy.1 \
+	man1/flux-cron.1 \
+	man1/flux-event.1 \
+	man1/flux-mini.1 \
+	man1/flux-version.1 \
+	man1/flux-jobs.1
 
 # These files are generated as clones of a primary page.
 # Sphinx handles this automatically if declared in the conf.py
 MAN1_FILES_SECONDARY = \
-	man/flux-setattr.1 \
-	man/flux-lsattr.1
+	man1/flux-setattr.1 \
+	man1/flux-lsattr.1
 
 
 MAN3_FILES = $(MAN3_FILES_PRIMARY) $(MAN3_FILES_SECONDARY)
 
 MAN3_FILES_PRIMARY = \
-	man/flux_shell_task_channel_subscribe.3 \
-	man/flux_shell_add_completion_ref.3 \
-	man/flux_shell_add_event_context.3 \
-	man/flux_shell_add_event_handler.3 \
-	man/flux_shell_aux_set.3 \
-	man/flux_shell_current_task.3 \
-	man/flux_shell_get_flux.3 \
-	man/flux_shell_get_info.3 \
-	man/flux_shell_getenv.3 \
-	man/flux_shell_getopt.3 \
-	man/flux_shell_killall.3 \
-	man/flux_shell_log.3 \
-	man/flux_shell_plugstack_call.3 \
-	man/flux_shell_rpc_pack.3 \
-	man/flux_shell_service_register.3 \
-	man/flux_shell_task_get_info.3 \
-	man/flux_shell_task_subprocess.3 \
-	man/flux_open.3 \
-	man/flux_send.3 \
-	man/flux_recv.3 \
-	man/flux_requeue.3 \
-	man/flux_aux_set.3 \
-	man/flux_flags_set.3 \
-	man/flux_fatal_set.3 \
-	man/flux_event_subscribe.3 \
-	man/flux_event_publish.3 \
-	man/flux_pollevents.3 \
-	man/flux_msg_encode.3 \
-	man/flux_rpc.3 \
-	man/flux_get_rank.3 \
-	man/flux_attr_get.3 \
-	man/flux_get_reactor.3 \
-	man/flux_reactor_create.3 \
-	man/flux_fd_watcher_create.3 \
-	man/flux_watcher_start.3 \
-	man/flux_zmq_watcher_create.3 \
-	man/flux_handle_watcher_create.3 \
-	man/flux_timer_watcher_create.3 \
-	man/flux_periodic_watcher_create.3 \
-	man/flux_idle_watcher_create.3 \
-	man/flux_msg_handler_create.3 \
-	man/flux_msg_cmp.3 \
-	man/flux_msg_handler_addvec.3 \
-	man/flux_child_watcher_create.3 \
-	man/flux_signal_watcher_create.3 \
-	man/flux_stat_watcher_create.3 \
-	man/flux_respond.3 \
-	man/flux_reactor_now.3 \
-	man/flux_request_decode.3 \
-	man/flux_event_decode.3 \
-	man/flux_response_decode.3 \
-	man/flux_request_encode.3 \
-	man/flux_content_load.3 \
-	man/flux_log.3 \
-	man/flux_future_get.3 \
-	man/flux_future_create.3 \
-	man/flux_future_wait_all_create.3 \
-	man/flux_future_and_then.3 \
-	man/flux_kvs_lookup.3 \
-	man/flux_kvs_commit.3 \
-	man/flux_kvs_txn_create.3 \
-	man/flux_kvs_namespace_create.3 \
-	man/flux_kvs_getroot.3 \
-	man/flux_kvs_copy.3 \
-	man/flux_core_version.3 \
-	man/idset_create.3 \
-	man/idset_encode.3
+	man3/flux_shell_task_channel_subscribe.3 \
+	man3/flux_shell_add_completion_ref.3 \
+	man3/flux_shell_add_event_context.3 \
+	man3/flux_shell_add_event_handler.3 \
+	man3/flux_shell_aux_set.3 \
+	man3/flux_shell_current_task.3 \
+	man3/flux_shell_get_flux.3 \
+	man3/flux_shell_get_info.3 \
+	man3/flux_shell_getenv.3 \
+	man3/flux_shell_getopt.3 \
+	man3/flux_shell_killall.3 \
+	man3/flux_shell_log.3 \
+	man3/flux_shell_plugstack_call.3 \
+	man3/flux_shell_rpc_pack.3 \
+	man3/flux_shell_service_register.3 \
+	man3/flux_shell_task_get_info.3 \
+	man3/flux_shell_task_subprocess.3 \
+	man3/flux_open.3 \
+	man3/flux_send.3 \
+	man3/flux_recv.3 \
+	man3/flux_requeue.3 \
+	man3/flux_aux_set.3 \
+	man3/flux_flags_set.3 \
+	man3/flux_fatal_set.3 \
+	man3/flux_event_subscribe.3 \
+	man3/flux_event_publish.3 \
+	man3/flux_pollevents.3 \
+	man3/flux_msg_encode.3 \
+	man3/flux_rpc.3 \
+	man3/flux_get_rank.3 \
+	man3/flux_attr_get.3 \
+	man3/flux_get_reactor.3 \
+	man3/flux_reactor_create.3 \
+	man3/flux_fd_watcher_create.3 \
+	man3/flux_watcher_start.3 \
+	man3/flux_zmq_watcher_create.3 \
+	man3/flux_handle_watcher_create.3 \
+	man3/flux_timer_watcher_create.3 \
+	man3/flux_periodic_watcher_create.3 \
+	man3/flux_idle_watcher_create.3 \
+	man3/flux_msg_handler_create.3 \
+	man3/flux_msg_cmp.3 \
+	man3/flux_msg_handler_addvec.3 \
+	man3/flux_child_watcher_create.3 \
+	man3/flux_signal_watcher_create.3 \
+	man3/flux_stat_watcher_create.3 \
+	man3/flux_respond.3 \
+	man3/flux_reactor_now.3 \
+	man3/flux_request_decode.3 \
+	man3/flux_event_decode.3 \
+	man3/flux_response_decode.3 \
+	man3/flux_request_encode.3 \
+	man3/flux_content_load.3 \
+	man3/flux_log.3 \
+	man3/flux_future_get.3 \
+	man3/flux_future_create.3 \
+	man3/flux_future_wait_all_create.3 \
+	man3/flux_future_and_then.3 \
+	man3/flux_kvs_lookup.3 \
+	man3/flux_kvs_commit.3 \
+	man3/flux_kvs_txn_create.3 \
+	man3/flux_kvs_namespace_create.3 \
+	man3/flux_kvs_getroot.3 \
+	man3/flux_kvs_copy.3 \
+	man3/flux_core_version.3 \
+	man3/idset_create.3 \
+	man3/idset_encode.3
 
 # These files are generated as clones of a primary page.
 # Sphinx handles this automatically if declared in the conf.py
 MAN3_FILES_SECONDARY = \
-	man/flux_shell_remove_completion_ref.3 \
-	man/flux_shell_aux_get.3 \
-	man/flux_shell_task_first.3 \
-	man/flux_shell_task_next.3 \
-	man/flux_shell_info_unpack.3 \
-	man/flux_shell_get_rank_info.3 \
-	man/flux_shell_rank_info_unpack.3 \
-	man/flux_shell_get_environ.3 \
-	man/flux_shell_setenvf.3 \
-	man/flux_shell_unsetenv.3 \
-	man/flux_shell_getopt_unpack.3 \
-	man/flux_shell_setopt.3 \
-	man/flux_shell_setopt_pack.3 \
-	man/flux_shell_err.3 \
-	man/flux_shell_fatal.3 \
-	man/flux_shell_log_setlevel.3 \
-	man/flux_shell_task_info_unpack.3 \
-	man/flux_shell_task_cmd.3 \
-	man/flux_clone.3 \
-	man/flux_close.3 \
-	man/flux_requeue_nocopy.3 \
-	man/flux_aux_get.3 \
-	man/flux_flags_unset.3 \
-	man/flux_flags_get.3 \
-	man/flux_fatal_error.3 \
-	man/FLUX_FATAL.3 \
-	man/flux_event_unsubscribe.3 \
-	man/flux_event_publish_pack.3 \
-	man/flux_event_publish_raw.3 \
-	man/flux_event_publish_get_seq.3 \
-	man/flux_pollfd.3 \
-	man/flux_msg_decode.3 \
-	man/flux_get_size.3 \
-	man/flux_attr_set.3 \
-	man/flux_set_reactor.3 \
-	man/flux_reactor_destroy.3 \
-	man/flux_reactor_run.3 \
-	man/flux_reactor_stop.3 \
-	man/flux_reactor_stop_error.3 \
-	man/flux_reactor_active_incref.3 \
-	man/flux_reactor_active_decref.3 \
-	man/flux_fd_watcher_get_fd.3 \
-	man/flux_watcher_stop.3 \
-	man/flux_watcher_destroy.3 \
-	man/flux_watcher_next_wakeup.3 \
-	man/flux_zmq_watcher_get_zsock.3 \
-	man/flux_handle_watcher_get_flux.3 \
-	man/flux_timer_watcher_reset.3 \
-	man/flux_periodic_watcher_reset.3 \
-	man/flux_prepare_watcher_create.3 \
-	man/flux_check_watcher_create.3 \
-	man/flux_msg_handler_destroy.3 \
-	man/flux_msg_handler_start.3 \
-	man/flux_msg_handler_stop.3 \
-	man/flux_msg_handler_delvec.3 \
-	man/flux_child_watcher_get_rpid.3 \
-	man/flux_child_watcher_get_rstatus.3 \
-	man/flux_signal_watcher_get_signum.3 \
-	man/flux_stat_watcher_get_rstat.3 \
-	man/flux_respond_raw.3 \
-	man/flux_respond_pack.3 \
-	man/flux_respond_error.3 \
-	man/flux_reactor_now_update.3 \
-	man/flux_request_unpack.3 \
-	man/flux_request_decode_raw.3 \
-	man/flux_event_unpack.3 \
-	man/flux_event_encode.3 \
-	man/flux_event_pack.3 \
-	man/flux_event_encode_raw.3 \
-	man/flux_event_decode_raw.3 \
-	man/flux_response_decode_raw.3 \
-	man/flux_response_decode_error.3 \
-	man/flux_request_encode_raw.3 \
-	man/flux_content_load_get.3 \
-	man/flux_content_store.3 \
-	man/flux_content_store_get.3 \
-	man/flux_vlog.3 \
-	man/flux_log_set_appname.3 \
-	man/flux_log_set_procid.3 \
-	man/flux_future_then.3 \
-	man/flux_future_wait_for.3 \
-	man/flux_future_reset.3 \
-	man/flux_future_destroy.3 \
-	man/flux_future_get.3 \
-	man/flux_future_fulfill.3 \
-	man/flux_future_fulfill_error.3 \
-	man/flux_future_fulfill_with.3 \
-	man/flux_future_aux_set.3 \
-	man/flux_future_aux_get.3 \
-	man/flux_future_set_flux.3 \
-	man/flux_future_get_flux.3 \
-	man/flux_future_wait_all_create.3 \
-	man/flux_future_wait_any_create.3 \
-	man/flux_future_push.3 \
-	man/flux_future_first_child.3 \
-	man/flux_future_next_child.3 \
-	man/flux_future_get_child.3 \
-	man/flux_future_and_then.3
-	man/flux_future_or_then.3 \
-	man/flux_future_continue.3 \
-	man/flux_future_continue_error.3 \
-	man/flux_rpc_pack.3 \
-	man/flux_rpc_raw.3 \
-	man/flux_rpc_message.3 \
-	man/flux_rpc_get.3 \
-	man/flux_rpc_get_unpack.3 \
-	man/flux_rpc_get_raw.3 \
-	man/flux_kvs_lookupat.3 \
-	man/flux_kvs_lookup_get.3 \
-	man/flux_kvs_lookup_get_unpack.3 \
-	man/flux_kvs_lookup_get_raw.3 \
-	man/flux_kvs_lookup_get_dir.3 \
-	man/flux_kvs_lookup_get_treeobj.3 \
-	man/flux_kvs_lookup_get_symlink.3 \
-	man/flux_kvs_getroot_get_treeobj.3 \
-	man/flux_kvs_getroot_get_blobref.3 \
-	man/flux_kvs_getroot_get_sequence.3 \
-	man/flux_kvs_getroot_get_owner.3 \
-	man/flux_kvs_getroot_cancel.3 \
-	man/flux_kvs_fence.3 \
-	man/flux_kvs_commit_get_treeobj.3 \
-	man/flux_kvs_commit_get_sequence.3 \
-	man/flux_kvs_txn_destroy.3 \
-	man/flux_kvs_txn_put.3 \
-	man/flux_kvs_txn_pack.3 \
-	man/flux_kvs_txn_vpack.3 \
-	man/flux_kvs_txn_mkdir.3 \
-	man/flux_kvs_txn_unlink.3 \
-	man/flux_kvs_txn_symlink.3 \
-	man/flux_kvs_txn_put_raw.3 \
-	man/flux_kvs_namespace_remove.3 \
-	man/flux_kvs_move.3 \
-	man/flux_core_version_string.3 \
-	man/idset_destroy.3 \
-	man/idset_decode.3 \
-	man/idset_ndecode.3 \
-	man/idset_format_map.3 \
-	man/idset_set.3 \
-	man/idset_clear.3 \
-	man/idset_first.3 \
-	man/idset_next.3 \
-	man/idset_count.3 \
-	man/idset_equal.3
+	man3/flux_shell_remove_completion_ref.3 \
+	man3/flux_shell_aux_get.3 \
+	man3/flux_shell_task_first.3 \
+	man3/flux_shell_task_next.3 \
+	man3/flux_shell_info_unpack.3 \
+	man3/flux_shell_get_rank_info.3 \
+	man3/flux_shell_rank_info_unpack.3 \
+	man3/flux_shell_get_environ.3 \
+	man3/flux_shell_setenvf.3 \
+	man3/flux_shell_unsetenv.3 \
+	man3/flux_shell_getopt_unpack.3 \
+	man3/flux_shell_setopt.3 \
+	man3/flux_shell_setopt_pack.3 \
+	man3/flux_shell_err.3 \
+	man3/flux_shell_fatal.3 \
+	man3/flux_shell_log_setlevel.3 \
+	man3/flux_shell_task_info_unpack.3 \
+	man3/flux_shell_task_cmd.3 \
+	man3/flux_clone.3 \
+	man3/flux_close.3 \
+	man3/flux_requeue_nocopy.3 \
+	man3/flux_aux_get.3 \
+	man3/flux_flags_unset.3 \
+	man3/flux_flags_get.3 \
+	man3/flux_fatal_error.3 \
+	man3/FLUX_FATAL.3 \
+	man3/flux_event_unsubscribe.3 \
+	man3/flux_event_publish_pack.3 \
+	man3/flux_event_publish_raw.3 \
+	man3/flux_event_publish_get_seq.3 \
+	man3/flux_pollfd.3 \
+	man3/flux_msg_decode.3 \
+	man3/flux_get_size.3 \
+	man3/flux_attr_set.3 \
+	man3/flux_set_reactor.3 \
+	man3/flux_reactor_destroy.3 \
+	man3/flux_reactor_run.3 \
+	man3/flux_reactor_stop.3 \
+	man3/flux_reactor_stop_error.3 \
+	man3/flux_reactor_active_incref.3 \
+	man3/flux_reactor_active_decref.3 \
+	man3/flux_fd_watcher_get_fd.3 \
+	man3/flux_watcher_stop.3 \
+	man3/flux_watcher_destroy.3 \
+	man3/flux_watcher_next_wakeup.3 \
+	man3/flux_zmq_watcher_get_zsock.3 \
+	man3/flux_handle_watcher_get_flux.3 \
+	man3/flux_timer_watcher_reset.3 \
+	man3/flux_periodic_watcher_reset.3 \
+	man3/flux_prepare_watcher_create.3 \
+	man3/flux_check_watcher_create.3 \
+	man3/flux_msg_handler_destroy.3 \
+	man3/flux_msg_handler_start.3 \
+	man3/flux_msg_handler_stop.3 \
+	man3/flux_msg_handler_delvec.3 \
+	man3/flux_child_watcher_get_rpid.3 \
+	man3/flux_child_watcher_get_rstatus.3 \
+	man3/flux_signal_watcher_get_signum.3 \
+	man3/flux_stat_watcher_get_rstat.3 \
+	man3/flux_respond_raw.3 \
+	man3/flux_respond_pack.3 \
+	man3/flux_respond_error.3 \
+	man3/flux_reactor_now_update.3 \
+	man3/flux_request_unpack.3 \
+	man3/flux_request_decode_raw.3 \
+	man3/flux_event_unpack.3 \
+	man3/flux_event_encode.3 \
+	man3/flux_event_pack.3 \
+	man3/flux_event_encode_raw.3 \
+	man3/flux_event_decode_raw.3 \
+	man3/flux_response_decode_raw.3 \
+	man3/flux_response_decode_error.3 \
+	man3/flux_request_encode_raw.3 \
+	man3/flux_content_load_get.3 \
+	man3/flux_content_store.3 \
+	man3/flux_content_store_get.3 \
+	man3/flux_vlog.3 \
+	man3/flux_log_set_appname.3 \
+	man3/flux_log_set_procid.3 \
+	man3/flux_future_then.3 \
+	man3/flux_future_wait_for.3 \
+	man3/flux_future_reset.3 \
+	man3/flux_future_destroy.3 \
+	man3/flux_future_fulfill.3 \
+	man3/flux_future_fulfill_error.3 \
+	man3/flux_future_fulfill_with.3 \
+	man3/flux_future_aux_set.3 \
+	man3/flux_future_aux_get.3 \
+	man3/flux_future_set_flux.3 \
+	man3/flux_future_get_flux.3 \
+	man3/flux_future_wait_any_create.3 \
+	man3/flux_future_push.3 \
+	man3/flux_future_first_child.3 \
+	man3/flux_future_next_child.3 \
+	man3/flux_future_get_child.3 \
+	man3/flux_future_or_then.3 \
+	man3/flux_future_continue.3 \
+	man3/flux_future_continue_error.3 \
+	man3/flux_rpc_pack.3 \
+	man3/flux_rpc_raw.3 \
+	man3/flux_rpc_message.3 \
+	man3/flux_rpc_get.3 \
+	man3/flux_rpc_get_unpack.3 \
+	man3/flux_rpc_get_raw.3 \
+	man3/flux_kvs_lookupat.3 \
+	man3/flux_kvs_lookup_get.3 \
+	man3/flux_kvs_lookup_get_unpack.3 \
+	man3/flux_kvs_lookup_get_raw.3 \
+	man3/flux_kvs_lookup_get_dir.3 \
+	man3/flux_kvs_lookup_get_treeobj.3 \
+	man3/flux_kvs_lookup_get_symlink.3 \
+	man3/flux_kvs_getroot_get_treeobj.3 \
+	man3/flux_kvs_getroot_get_blobref.3 \
+	man3/flux_kvs_getroot_get_sequence.3 \
+	man3/flux_kvs_getroot_get_owner.3 \
+	man3/flux_kvs_getroot_cancel.3 \
+	man3/flux_kvs_fence.3 \
+	man3/flux_kvs_commit_get_treeobj.3 \
+	man3/flux_kvs_commit_get_sequence.3 \
+	man3/flux_kvs_txn_destroy.3 \
+	man3/flux_kvs_txn_put.3 \
+	man3/flux_kvs_txn_pack.3 \
+	man3/flux_kvs_txn_vpack.3 \
+	man3/flux_kvs_txn_mkdir.3 \
+	man3/flux_kvs_txn_unlink.3 \
+	man3/flux_kvs_txn_symlink.3 \
+	man3/flux_kvs_txn_put_raw.3 \
+	man3/flux_kvs_txn_put_treeobj.3 \
+	man3/flux_kvs_namespace_remove.3 \
+	man3/flux_kvs_move.3 \
+	man3/flux_core_version_string.3 \
+	man3/idset_destroy.3 \
+	man3/idset_decode.3 \
+	man3/idset_ndecode.3 \
+	man3/idset_format_map.3 \
+	man3/idset_set.3 \
+	man3/idset_clear.3 \
+	man3/idset_first.3 \
+	man3/idset_next.3 \
+	man3/idset_count.3 \
+	man3/idset_equal.3
 
 MAN5_FILES = $(MAN5_FILES_PRIMARY)
 
 MAN5_FILES_PRIMARY = \
-	man/flux-config-bootstrap.5
+	man5/flux-config-bootstrap.5
 
 MAN7_FILES = $(MAN7_FILES_PRIMARY)
 
 MAN7_FILES_PRIMARY = \
-	man/flux-broker-attributes.7
+	man7/flux-broker-attributes.7
 
 
-RST_FILES  = $(MAN1_FILES_PRIMARY:man/%.1=man1/%.rst)
-	$(MAN3_FILES_PRIMARY:man/%.3=man3/%.rst) \
-	$(MAN5_FILES_PRIMARY:man/%.5=man5/%.rst) \
-	$(MAN7_FILES_PRIMARY:man/%.7=man7/%.rst)
+RST_FILES  = $(MAN1_FILES_PRIMARY:.1=.rst)
+	$(MAN3_FILES_PRIMARY:.3=.rst) \
+	$(MAN5_FILES_PRIMARY:.5=.rst) \
+	$(MAN7_FILES_PRIMARY:.7=.rst)
 
 if ENABLE_DOCS
 dist_man_MANS = $(MAN1_FILES) $(MAN3_FILES) $(MAN5_FILES) $(MAN7_FILES)
@@ -282,16 +280,38 @@ endif
 
 SUFFIXES = .rst .1 .3 .5 .7
 
-$(MAN_FILES): $(RST_FILES)
-	$(AM_V_GEN) $(PYTHON) -m sphinx -M man ./ ./
+sphinx_verbose = $(sphinx_verbose_$(V))
+sphinx_verbose_ = $(sphinx_verbose_$(AM_DEFAULT_VERBOSITY))
+sphinx_verbose_0 = @echo "  BUILD     manpages";
+
+STDERR_DEVNULL = $(stderr_devnull_$(V))
+stderr_devnull_ =  $(stderr_devnull_$(AM_DEFAULT_VERBOSITY))
+stderr_devnull_0 = >/dev/null 2>&1
+
+$(MAN_FILES): conf.py $(RST_FILES)
+	$(sphinx_verbose)$(PYTHON) -m sphinx -M man $(srcdir) ./ $(STDERR_DEVNULL)
+	@echo "  MV        manpages"; \
+	for sec in 1 3 5 7; do \
+	  $(MKDIR_P) man$$sec && \
+	  mv -f $(abs_builddir)/man/*.$$sec man$$sec/; \
+	done
 
 EXTRA_DIST = \
+	conf.py \
 	$(RST_FILES) \
 	man1/NODESET.rst \
 	man3/JSON_PACK.rst \
 	man3/JSON_UNPACK.rst
 
-CLEANFILES = $(MAN_FILES)
+CLEANFILES = \
+	$(MAN_FILES)
+
+clean-local:
+	-rm -rf man
+	-rmdir man1 man3 man5 man7
+
+distclean-local:
+	-rm -rf doctrees
 
 # test programs from man3
 AM_CFLAGS = \
@@ -302,15 +322,15 @@ AM_LDFLAGS = \
 	$(CODE_COVERAGE_LIBS)
 
 AM_CPPFLAGS = \
-        -I$(top_srcdir) \
+	-I$(top_srcdir) \
 	-I$(top_srcdir)/src/include \
 	-I$(top_builddir)/src/common/libflux \
-        $(ZMQ_CFLAGS) $(LIBUUID_CFLAGS)
+	$(ZMQ_CFLAGS) $(LIBUUID_CFLAGS)
 
 LDADD = \
-        $(top_builddir)/src/common/libflux-internal.la \
-        $(top_builddir)/src/common/libflux-core.la \
-        $(ZMQ_LIBS) $(LIBUUID_CFLAGS) $(LIBPTHREAD)
+	$(top_builddir)/src/common/libflux-internal.la \
+	$(top_builddir)/src/common/libflux-core.la \
+	$(ZMQ_LIBS) $(LIBUUID_CFLAGS) $(LIBPTHREAD)
 
 check_PROGRAMS = \
 	man3/topen \

```

grondo commented:
@gonsie, I apologize I didn't pull from your branch before working on changes above. I've done that now and collapsed my changes with yours in 94d254199692f01ad25b03c8583099a25b52bfeb on my branch. Sorry if that got a bit messy. In the end I'd be fine if most of this intermediate work is all squashed together anyway.

Anyway, the MANPATH is working when running flux under build tree now:
```
ƒ(s=1,d=0,builddir) grondo@asp:~/git/f.git/build/doc$ make
Making all in .
make[1]: Entering directory '/home/grondo/git/f.git/build/doc'
  BUILD     manpages
  MV        manpages
make[1]: Leaving directory '/home/grondo/git/f.git/build/doc'
Making all in test
make[1]: Entering directory '/home/grondo/git/f.git/build/doc/test'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/grondo/git/f.git/build/doc/test'
ƒ(s=1,d=0,builddir) grondo@asp:~/git/f.git/build/doc$ man -w flux
/home/grondo/git/f.git/build/doc/man1/flux.1

```

grondo commented:
> also the build tree passes make distcheck!

Well I spoke too soon, the distcheck passes only because make doesn't fail when `gen-cmdhelp.pl` does.

gonsie commented:
@grondo Thanks for the commit! I have a gen-cmdhelp replacement ready. I'll commit it and your changes tomorrow and hopefully this PR will be ready!

grondo commented:
> Thanks for the commit! I have a gen-cmdhelp replacement ready. I'll commit it and your changes tomorrow and hopefully this PR will be ready!

Great! I'm willing to help in any way you'd like! Would be nice to get this merged!

gonsie commented:
@grondo Can you double check my latest commit, especially the makefile changes? I removed the `gen-cmdhelp.pl` as the core.json file is now generated by sphinx (code in conf.py).

Other than that, I think I’ve addressed everything else and this should be ready to merge!

grondo commented:
>  I removed the gen-cmdhelp.pl as the core.json file is now generated by sphinx (code in conf.py).

Great improvement!

I'll check it out after lunch.

This PR needs to be rebased on current master and conflicts resolved. I'm perfectly happy to do that if you don't mind,  @gonsie.

gonsie commented:
> This PR needs to be rebased on current master and conflicts resolved. I'm perfectly happy to do that if you don't mind

@grondo If you could rebase that would be awesome!

grondo commented:
> @grondo If you could rebase that would be awesome!

Sure! I can take it from here. I'll force-push to your branch if that is ok.

gonsie commented:
Yep, do what ever you need.

grondo commented:
Ok! I just rebased, squashed all the intermediate work down to 3 commits, and forced a push.

@gonsie, I had to move the cmdhelp generator out of sphinx's `conf.py`, there was some strangeness with paths there that I couldn't quite get to the bottom of, so I moved it back to `etc/gen-cmdhelp.py`, but the script still reads `conf.py` to get the list of section 1 manpages and command names and description.

This is working on my local system with `make distcheck` so I think we are ready to merge!

The last thing to do is to generate the new docker images so this branch can build in Travis, but once I do that all other PRs may fail until this PR is merged. (They might not though, perhaps just docs will be disabled). Just wanted to ensure everyone is aware of that before I go ahead...

grondo commented:
Ah, one more fix needed here. If sphinx is missing `./configure` fails. I'll try to reestablish the previous behavior of just disabling doc generation.

grondo commented:
Ok, replaced a `AC_MSG_ERROR` with `AC_MSG_WARN`, and added back the other WITH_DOCS logic to `configure.ac`. This should be ready now! I'll go ahead an build new docker images and test just to be sure.

gonsie commented:
👏🏼👏🏼👏🏼thanks @grondo !!

grondo commented:
> thanks @grondo !!

We would have never been able to get this done without you @gonsie! Thanks so much for all the tedious work!

garlick commented:
Thanks everybody!  I'm excited to see how these look.

grondo commented:
Have to rebase again since #3068 went in.

SteVwonder commented:
Thanks @gonsie and @grondo for all the effort on this. 🥳 👏 🙌 🍾 🥂 🍰 

> I'm excited to see how these look.

Same! It is going to be 🔥 to have our manpages up on readthedocs.

grondo commented:
In the process of pushing up the new docker images. Then this is officially ready for merge.

grondo commented:
CI is now all green. @garlick gave an approval, @SteVwonder want to take a look too or shall we set MWP?

grondo commented:
Sorry @garlick, we cross posted. I'm ready to set MWP. We can fix up any fallout as we go along.

SteVwonder commented:
MWP is fine with me.  I agree, let's fix stuff in a follow up PR if anything is broken.  Kinda confused why we aren't seeing the RTD's CI builder check thingy.  I thought we added the webhook.  I'll poke at that.

grondo commented:
Oh, yeah I meant to ask about that. I didn't see anything that pushes the built docs to RTD. (Though I didn't look carefully, was mainly concentrating on getting the conversion done so we don't conflict with future manpage updates.)

grondo commented:
MWP set!

grondo commented:
Thanks @gonsie and @SteVwonder!! :clap:

SteVwonder commented:
YAY!!!!!!! 🥳 

Per the RTD CI check, the webhook was working just fine. The issue was that the builder for PRs wasn't enabled in the Advanced Settings: https://docs.readthedocs.io/en/stable/guides/autobuild-docs-for-pull-requests.html.  I flicked that on, hopefully it works now. 🤞 

I also took the liberty of updating some RTD settings like where it looks for the `conf.py` and the `requirements.txt`.  FWIW, if we need to tweak those setting in the future, we can do so with Yet Another Dotfile at the Top-Level of the Repo (YADTLR™️): https://docs.readthedocs.io/en/stable/config-file/v2.html

grondo commented:
Great, thanks! It is working

https://flux-framework.readthedocs.io/projects/flux-core/en/latest/man1/flux-jobs.html

One thing I notice is that on the left sidebar all the commands are listed in caps, e.g. `FLUX-JOBS(1)`. Just something minor to fixup later.

garlick commented:
Looks great!  One other nice to have would be inline manpage references working as links, but probably for another time.  Nice work everybody!
